### ISSUE TITLE
[TensorGraph] predict_on_generator issues

### ISSUE THREAD

**Issue by patrickhop** [deepchem/deepchem#579]
I'm having issues getting predictions out of tensorgraph models using the predict_on_generator. The outputs seem to be somewhat random, and don't match the R2 results of evaluate_generator.

Am I using it wrong, or is there something with the internals here?

```
from __future__ import print_function
from __future__ import division
from __future__ import unicode_literals

import numpy as np
import six

from deepchem.models.tensorgraph import TensorGraph
from deepchem.metrics import to_one_hot

from deepchem.feat.mol_graphs import ConvMol
from deepchem.models.tensorgraph.layers import Input, GraphConv, BatchNorm, GraphPool, Dense, GraphGather, \
  SoftMax, SoftMaxCrossEntropy, Concat, WeightedError, Label, Weights, Feature, L2Loss

np.random.seed(123)
import tensorflow as tf

tf.set_random_seed(123)
import deepchem as dc
import donkey
import models

DATASET = 'data/az_logd.csv'
SPLITTER = 'scaffold'
ALPHA = 1e-3 # not supported in tens_graph yet
BATCH_SIZE = 48
MAX_EPOCH = 4

tasks, datasets, transformers = donkey.load_dataset(DATASET,
                                                    ['azlogd'],
                                                    featurizer='GraphConv',
                                                    split=SPLITTER,
                                                    frac_train=0.80,
                                                    frac_val=0.10,
                                                    frac_test=0.10)
train_dataset, val_dataset, test_dataset = datasets

metric = dc.metrics.Metric(dc.metrics.pearson_r2_score, mode="regression")
model, generator, labels, task_weights = models.gcdnn(BATCH_SIZE, tasks)

# flag best model
for i in xrange(0, MAX_EPOCH):
  model.fit_generator(generator(train_dataset, BATCH_SIZE, epochs=4))
  print('\ncompleted [%d] epochs' % (i+1))
  valid_scores = model.evaluate_generator(
      generator(val_dataset, BATCH_SIZE), [metric],
      transformers,
      labels)

from sklearn.metrics import r2_score

out = model.predict_on_generator(
      generator(val_dataset, 48),
      transformers)

y = val_dataset.y
ids = val_dataset.ids
y = dc.trans.undo_transforms(val_dataset.y, transformers)
y = map(lambda x: x[0], y)
out = map(lambda x: x[0][0], out)
print(len(out))
print(len(y))
print(out[:10])
print(y[:10])
print(ids[:10])

print(r2_score(y[:200], out[:200]))
```

**Comment by rbharath** [comment]
What's the precise error you're seeing? From first glance, usage looks right to me.

Also, would help if you could post a minimal failing example (the smaller the better)

**Comment by patrickhop** [comment]
https://www.dropbox.com/s/3gllsh9m9vbiumx/Screen%20Shot%202017-05-30%20at%208.19.03%20PM.png?dl=0

It seems that the predicted has little to do with the labels. Sklearn is giving me an R2 of -0.236 for this one, whereas evaluate_generator gives me like .45

Are there any tensorgraph regression examples I can use to construct a minimum failing example? I'm happy to put the effort in, I just don't really know how.

**Comment by rbharath** [comment]
The key to making a minimal failing example is to reduce the moving parts. For example, we can't repro the failure since it's using a private dataset. Does a simple synthetic dataset cause the same failure? Also, why use a for loop for epochs when fit_generator already has num epochs as an argument. Removing unneeded bits of code will help us understand if there really is an error, and if so, how we can debug it.

**Comment by patrickhop** [comment]
Resolved. Closing.
